### DISCUSSION TITLE
When are Loops and List Comprehensions valid under jitted functions?

### DISCUSSION THREAD

**Discussion by aniquetahir** [jax-ml/jax#13449] [Q&A] [answered]
I was recently working on a jax function where I needed to recurse over a `PyTree` and I found that everything I knew about jax was a lie and I can actually use list comprehensions and python loops in jittable functions.
e.g.
```python3
@functools.partial(jit, static_argnums=(1,))
def mult_two(x, n):
    for i in range(n):
        x = x * 2
    return x
```
Now I'm wondering what are the cases where python loops and list comprehensions are jit compileable?

**Comment by jakevdp** [comment, marked as answer]
Thanks for the question: python loops will be valid under JIT when their length is static. So for example this will result in an error because `n` is a traced quantity:
```python
@jit
def mult_two(x, n):
    for i in range(n):
        x = x * 2
    return x
```
Whereas this will be valid because `n` is static (i.e. a compile-time constant):
```python
@functools.partial(jit, static_argnums=(1,))
def mult_two(x, n):
    for i in range(n):
        x = x * 2
    return x
```
Similarly, this would be valid because the loop value is a constant:
```python
@jit
def mult_two(x):
    for i in range(5):
        x = x * 2
    return x
```
You can read more on this topic at [Sharp Bits: Python Control Flow and JIT](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#python-control-flow-jit).
